crates/legacy/src/legacy_testing/mod.rs

```diff
@@ -0,0 +1,278 @@
+        broadcast::<MessageType<TestTypes>>(channel_capacity);
+    let (senders, _receivers) = broadcast::<MessageType<TestTypes>>(channel_capacity);
+    let (tx_sender, tx_receiver) =
+        broadcast::<Arc<ReceivedTransaction<TestTypes>>>(channel_capacity);
+
+    let genesis_vid_commitment = vid_commitment(&[], num_storage_nodes);
+    let genesis_builder_commitment = BuilderCommitment::from_bytes([]);
+
+    // instantiate the global state
+    let global_state = Arc::new(RwLock::new(GlobalState::<TestTypes>::new(
+        bootstrap_sender,
+        tx_sender.clone(),
+        genesis_vid_commitment,
+        ViewNumber::genesis(),
+        ViewNumber::genesis(),
+        10,
+    )));
+
+    // instantiate the bootstrap builder state
+    let builder_state = BuilderState::new(
+        ParentBlockReferences {
+            view_number: ViewNumber::new(0),
+            vid_commitment: genesis_vid_commitment,
+            leaf_commit: Commitment::<Leaf<TestTypes>>::default_commitment_no_preimage(),
+            builder_commitment: genesis_builder_commitment,
+        },
+        decide_receiver.clone(),
+        da_receiver.clone(),
+        quorum_proposal_receiver.clone(),
+        bootstrap_receiver,
+        tx_receiver,
+        VecDeque::new(),
+        global_state.clone(),
+        NonZeroUsize::new(num_storage_nodes).unwrap(),
+        Duration::from_millis(100),
+        1,
+        Arc::new(TestInstanceState::default()),
+        Duration::from_millis(100),
+        Arc::new(TestValidatedState::default()),
+    );
+
+    (senders, global_state, builder_state)
+}
+
+/// get transactions submitted in previous rounds, [] for genesis
+/// and simulate the block built from those
+pub async fn calc_proposal_msg(
+    num_storage_nodes: usize,
+    round: usize,
+    prev_quorum_proposal: Option<QuorumProposal<TestTypes>>,
+    transactions: Vec<TestTransaction>,
+) -> (
+    QuorumProposal<TestTypes>,
+    MessageType<TestTypes>,
+    MessageType<TestTypes>,
+    BuilderStateId<TestTypes>,
+) {
+    // get transactions submitted in previous rounds, [] for genesis
+    // and simulate the block built from those
+    let num_transactions = transactions.len() as u64;
+    let encoded_transactions = TestTransaction::encode(&transactions);
+    let block_payload = TestBlockPayload { transactions };
+    let block_vid_commitment = vid_commitment(&encoded_transactions, num_storage_nodes);
+    let metadata = TestMetadata { num_transactions };
+    let block_builder_commitment =
+        <TestBlockPayload as BlockPayload<TestTypes>>::builder_commitment(
+            &block_payload,
+            &metadata,
+        );
+
+    // generate key for leader of this round
+    let seed = [round as u8; 32];
+    let (pub_key, private_key) = BLSPubKey::generated_from_seed_indexed(seed, round as u64);
+
+    // Prepare the DA proposal message
+    let da_proposal_message: DaProposalMessage<TestTypes> = {
```

Comment:
Is this separate scope needed? I see that it's shadowing `seed` and keys, but it looks like it shadows them with exact same values

Reply:
Good catch! Didn't notice that when I put them together. Removed in [75165e3](https://github.com/EspressoSystems/marketplace-builder-core/pull/176/commits/75165e35aa792efb87671ef150cb5af00ddc87cd)